src/datachain/data_storage/warehouse.py

```diff
@@ -232,7 +232,10 @@ def dataset_select_paginated(
                 if limit < page_size:
                     paginated_query = paginated_query.limit(None).limit(limit)
 
-            results = self.dataset_rows_select(paginated_query.offset(offset))
+            # Ensure we're using a thread-local connection
+            with self.clone() as wh:
```

Comment:
I guess this `clone()` should be outside the loop? Otherwise, we'll not be reusing the connection.